### PR TITLE
EVG-19750: fix time-series-update job race condition when there is both user-submitted rollups and FTDC rollups

### PR DESCRIPTION
**Pull request by julianedwards**
JIRA: https://jira.mongodb.org/browse/EVG-19750

- Only enqueue the time-series-update job once for a performance result upon the final call to `CloseMetrics`
- Add the `measurements` to the scope for each job
- Add more better testing to the perf results RPC service code

### CODE REVIEW

File: units/time_series_update.go [hunk 1]
```diff
@@ -57,7 +58,7 @@ func NewUpdateTimeSeriesJob(series model.UnanalyzedPerformanceSeries) amboy.Job
 		series.Arguments,
 	)
 	j.SetID(fmt.Sprintf("%s.%s", baseID, time.Now().UTC()))
-	j.SetScopes([]string{baseID})
+	j.SetScopes([]string{baseID, strings.Join(series.Measurements, ",")})
```
Comment (anjani-bhat) [inline units/time_series_update.go:61]: Is there any limit to how large this string can be? Joining all metrics would probably result in a very large string.

File: rpc/internal/perf_service.go [hunk 2]
```diff
@@ -228,21 +215,28 @@ func (srv *perfService) CloseMetrics(ctx context.Context, end *MetricsSeriesEnd)
...
+	// Only enqueue a new update time series job for mainline results if
+	// there are user-submitted rollups AND no event data (FTDC artifact).
+	// In the latter case, the FTDC rollups job will take care of
+	// enqueueing the update time series job.
+	if record.Info.Mainline && len(record.Rollups.Stats) > 0 && !hasEventData {
+		processingJob := units.NewUpdateTimeSeriesJob(record.CreateUnanalyzedSeries())
+		if err = amboy.EnqueueUniqueJob(ctx, srv.env.GetRemoteQueue(), processingJob); err != nil {
+			return nil, newRPCError(codes.Internal, errors.Wrapf(err, "creating signal processing job for perf result '%s'", record.ID))
+		}
 	}
```
Comment (anjani-bhat) [inline rpc/internal/perf_service.go:232]: Why is this moved from the CreateMetrics function to CloseMetrics function?

File: rpc/internal/perf_service.go [hunk 3]
```diff
@@ -228,21 +215,28 @@ func (srv *perfService) CloseMetrics(ctx context.Context, end *MetricsSeriesEnd)
...
+	}
 
-func (srv *perfService) addArtifacts(ctx context.Context, record *model.PerformanceResult, artifacts []*ArtifactInfo) error {
-	for _, a := range artifacts {
-		record.Artifacts = append(record.Artifacts, *a.Export())
+	// Only enqueue a new update time series job for mainline results if
+	// there are user-submitted rollups AND no event data (FTDC artifact).
+	// In the latter case, the FTDC rollups job will take care of
+	// enqueueing the update time series job.
+	if record.Info.Mainline && len(record.Rollups.Stats) > 0 && !hasEventData {
```
Comment (johndaniels) [inline rpc/internal/perf_service.go:227]: Just to confirm, this is a deliberate decision to only include Mainline since that's all that the Signal Processing Service handles, right?

File: rpc/internal/perf_service.go [hunk 4]
```diff
@@ -228,21 +215,28 @@ func (srv *perfService) CloseMetrics(ctx context.Context, end *MetricsSeriesEnd)
...
+		}
 	}
 
-	return errors.Wrap(srv.addFTDCRollupsJob(ctx, record.ID, record.Artifacts), "creating FTDC rollups job")
+	resp.Success = true
+	return resp, nil
 }
 
-func (srv *perfService) addFTDCRollupsJob(ctx context.Context, id string, artifacts []model.ArtifactInfo) error {
+func (srv *perfService) addFTDCRollupsJob(ctx context.Context, id string, artifacts []model.ArtifactInfo) (bool, error) {
```
Comment (johndaniels) [inline rpc/internal/perf_service.go:238]: Probably worth a doc comment, since it's not at all obvious what the bool here is.

File: rpc/internal/perf_service_test.go [hunk 5]
```diff
@@ -345,8 +384,8 @@ func TestAttachResultData(t *testing.T) {
 			defer func() {
 				require.NoError(t, tearDownEnv(env, false))
 			}()
-			port := getPort()
 
+			port := getPort()
```
Comment (johndaniels) [inline rpc/internal/perf_service_test.go:388]: ?

File: units/time_series_update.go (hunk 1 above)
Comment (johndaniels) [inline units/time_series_update.go:61]: What's the purpose of this change?

Comment (johndaniels) [reply units/time_series_update.go:61]: Does it matter that series.Measurements might have variable order?

File: rpc/internal/perf_service.go [hunk 6]
```diff
@@ -96,8 +83,8 @@ func (srv *perfService) AttachArtifacts(ctx context.Context, artifactData *Artif
 	resp := &MetricsResponse{}
 	resp.Id = record.ID
 
-	if err := srv.addArtifacts(ctx, record, artifactData.Artifacts); err != nil {
-		return resp, err
+	for _, a := range artifactData.Artifacts {
```
Comment (johndaniels) [inline rpc/internal/perf_service.go:86]: It's fine for AttachArtifacts to not send updates now, since it will all be sent together in CloseMetrics. Right?

**Review by johndaniels — COMMENTED** [review]
Mostly just want to make sure I understand the series.Measurements scope change.

File: units/time_series_update.go (hunk 1 above)
Comment (johndaniels) [reply units/time_series_update.go:61]: FWIW it's already in an array in memory  (series.Measurements), so this is probably fine from a memory usage perspective.

File: rpc/internal/perf_service_test.go (hunk 5 above)
Comment (julianedwards) [reply rpc/internal/perf_service_test.go:388]: Just moved it down a line lol. It felt like it belonged with this code block.

File: rpc/internal/perf_service.go [hunk 7]
```diff
@@ -251,19 +245,19 @@ func (srv *perfService) addFTDCRollupsJob(ctx context.Context, id string, artifa
 		}
 
 		if hasEventData {
-			return newRPCError(codes.InvalidArgument, errors.New("cannot have more than one raw events artifact"))
+			return false, newRPCError(codes.InvalidArgument, errors.New("cannot have more than one raw events artifact"))
```
[the remark this answers is not in the document]
Comment (julianedwards) [reply rpc/internal/perf_service.go:248]: `hasEventData` just means that it has a time series FTDC artifact, what is considered a`RawEvents` artifact schema.

File: units/time_series_update.go (hunk 1 above)
Comment (julianedwards) [reply units/time_series_update.go:61]: Yeah this is fine!

Comment (julianedwards) [reply units/time_series_update.go:61]: This is to prevent race conditions where new rollups are processed async from the same perf result but with different sets of measurements (usually one will be a subset of the other). If this happens, the job with the fuller set of measurements may not run since the scope is less selective without adding the measurements to it and the `analysis.processed_at` field on the perf result document would get set to a time greater than that of the `rollups.processed_at` field preventing the backfill aggregation from picking up the perf result. 

Adding this to the job scope enable two jobs with the same base ID but different measurements to get enqueued simultaneously—we will do some extra work by sending some measurements to SPS twice, but that is better than not sending them at all. I also mitigated the duplicate sending in the perf results RPC service code by enqueueing the FTDC rollups job/time-series-updated job when `CloseMetrics` is called.

Comment (julianedwards) [reply units/time_series_update.go:61]: The measurements are stored as arrays in the DB so order wouldn't be variable.

File: rpc/internal/perf_service.go (hunk 3 above)
Comment (julianedwards) [reply rpc/internal/perf_service.go:227]: Yes, the existing behavior already does this. I only added the last condition which is to check if the FTDC job was already enqueued.

File: rpc/internal/perf_service.go (hunk 2 above)
Comment (julianedwards) [reply rpc/internal/perf_service.go:232]: To ensure we only send it once per result and don't send duplicate notifications to SPS.

File: rpc/internal/perf_service.go (hunk 6 above)
Comment (julianedwards) [reply rpc/internal/perf_service.go:86]: Exactly! We only want to do this once to prevent duplicating the work here.

File: rpc/internal/perf_service.go (hunk 4 above)
Comment (julianedwards) [reply rpc/internal/perf_service.go:241]: Done!

File: rpc/internal/perf_service_test.go (hunk 5 above)
Comment (johndaniels) [reply rpc/internal/perf_service_test.go:388]: Ok so it was intentional, good.

File: units/time_series_update.go (hunk 1 above)
Comment (johndaniels) [reply units/time_series_update.go:61]: Thanks for the explanation.

**Review by johndaniels — APPROVED** [review]

**Review by anjani-bhat — APPROVED** [review]
